src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -338,7 +338,7 @@ private void EnsureUniqueLockFileTargetLibraries(LockFileTarget lockFileTarget)
 
             foreach (LockFileTargetLibrary library in libraries)
             {
-                var libraryKey = library.Name + " " + library.Version;
```

Comment:
The scope of the `libraryKey` is local to the method. I am not sure how adding a space between library name and version makes any difference. Hence removed it which reduced string allocations by around 2 MB for orchardcore solution restore scenrario.

Reply:
Maybe a custom comparer would be even better?

Reply:
It has a theoretical bug, although it's low risk.  Consider these two packages:

|Package Id|Version|
|---|---|
|My.Thing|21.2.3|
|MyThing2|1.2.3|

With the old code they will have different keys. With your change they will have the same key leading to bugs if there's even a single customer who hits it.

Reply:
@zivkan is right.  A custom comparer would remove the entire allocation of a string to calculate a key without much overhead.  I used a similar pattern here: 
https://github.com/NuGet/NuGet.Client/pull/4380/files#diff-9cbed86901743d01d5ef0ea288bfa1ad64f3594482d837ff241225098467e133R217

It has a static singleton so only one instance is ever created and it just compares the properties I wanted for the key.

Reply:
Thanks @zivkan for catching that scenario which I didn't think of while making that code change. 

@jeffkl Great suggestion. I will modify the code to use `LockFileTargetLibrary` as dictionary key along with a custom comparer to compare package id and version. If my understanding is correct, any attempt to add, update, lookup from the dictionary will result in a call to [`GetHashCode()`](https://github.com/NuGet/NuGet.Client/blob/e050b00c3683d3ecd18a8885ef464ec32b1e86b7/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs#L89). The said method is doing a lot of stuff (so many foreach loops). I am currently unsure if that increases method execution time. I will check the perfview trace and post it here before pushing the commit.

Reply:
> any attempt to add, update, lookup from the dictionary will result in a call to GetHashCode()

Yes but this is happening at the moment anyway since the key is a string.  The difference will be your custom comparer will get the hashcode for two strings and combine them.  That operation shouldn't be too bad and is probably still better than creating the string just to get a hash code anyway.

Reply:
I added a custom comparer in the latest commit and observed that string allocations reduced even further. Now the allocations have reduced from `18.5 MB to 7 MB` for OrchardCore solution restore. //cc @jeffkl